llama/generation.py

```diff
@@ -102,68 +104,124 @@ def build(
         max_gen_len: int,
         temperature: float = 0.6,
         top_p: float = 0.9,
-        logprobs: bool = False,
-        echo: bool = False,
-    ) -> Tuple[List[List[int]], Optional[List[List[float]]]]:
+        logprobs: bool = False,#NEW LINE
+        echo: bool = False,#NEW LINE
+    ) -> Tuple[List[List[int]], Optional[List[List[float]]]]: #NEW OUTPUT FORMAT
         params = self.model.params
-        bsz = len(prompt_tokens)
+        bsz = len(prompts)
+        #bsz = len(prompt_tokens)
         assert bsz <= params.max_batch_size, (bsz, params.max_batch_size)
 
+        prompt_tokens = [self.tokenizer.encode(x, bos=True, eos=False) for x in prompts]
+
+        # TODO: check if we need to optimize this block - llama1 has a simpler logic
         min_prompt_len = min(len(t) for t in prompt_tokens)
         max_prompt_len = max(len(t) for t in prompt_tokens)
         assert max_prompt_len <= params.max_seq_len
         total_len = min(params.max_seq_len, max_gen_len + max_prompt_len)
 
         pad_id = self.tokenizer.pad_id
-        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long, device="cuda")
+        tokens = torch.full((bsz, total_len), pad_id, dtype=torch.long)
         for k, t in enumerate(prompt_tokens):
-            tokens[k, : len(t)] = torch.tensor(t, dtype=torch.long, device="cuda")
-        if logprobs:
+            tokens[k, : len(t)] = torch.tensor(t, dtype=torch.long)
+        if logprobs: #TODO: NEW CODE - optimize this block
             token_logprobs = torch.zeros_like(tokens, dtype=torch.float)
+            token_logprobs.to(device)
 
-        prev_pos = 0
-        eos_reached = torch.tensor([False] * bsz, device="cuda")
+        device = xm.xla_device()
+        tokens = tokens.to(device)
+
+        start_pos = 1
+        cur_pos_tensor = torch.tensor(start_pos).to(device)
+        input_pos_tensor = torch.arange(0, start_pos).to(device)
+        output_pos_tensor = cur_pos_tensor - 1
+        input_tokens = tokens.index_select(1, input_pos_tensor)
+        cache_kvs = self.model.cache_kvs #TODO: revisit the cache implementation between the two models
+
+        prev_pos = 0 #TODO: drop this line
+        eos_reached = torch.tensor([False] * bsz, device=device)
         input_text_mask = tokens != pad_id
-        for cur_pos in range(min_prompt_len, total_len):
-            logits = self.model.forward(tokens[:, prev_pos:cur_pos], prev_pos)
-            if logprobs:
-                token_logprobs[:, prev_pos + 1 : cur_pos + 1] = -F.cross_entropy(
-                    input=logits.transpose(1, 2),
-                    target=tokens[:, prev_pos + 1 : cur_pos + 1],
-                    reduction="none",
-                    ignore_index=pad_id,
+        xm.mark_step(wait=True)
+
+        decoding_start_time = time.time()
+        for cur_pos in range(min_prompt_len, total_len): #TODO: drop dependency on cur_pos. min_prompt_len used to be start_pos
```

Comment:
`cur_pos` should match `cur_pos_tensor`.

Reply:
replaced this loop with what we had in llama1